proxystore/serialize.py

```diff
@@ -36,30 +106,16 @@ def serialize(obj: Any) -> bytes:
             serializers. Cloudpickle is the last resort, so this error will
             typically be raised from a cloudpickle error.
     """
-    if isinstance(obj, bytes):
-        identifier = b'01\n'
-    elif isinstance(obj, str):
-        identifier = b'02\n'
-        obj = obj.encode()
-    else:
-        # Use cloudpickle if pickle fails
-        try:
-            identifier = b'03\n'
-            # Pickle protocol 5 is available in Python 3.8 and later
-            obj = pickle.dumps(obj, protocol=5)
-        except Exception:
-            identifier = b'04\n'
+    for identifier, serializer in SERIALIZERS.items():
+        if serializer.serializable(obj):
             try:
-                obj = cloudpickle.dumps(obj)
-            except Exception as e:
-                raise SerializationError(
-                    f'Object of type {type(obj)} is not serializable.',
-                ) from e
-
-    assert isinstance(identifier, bytes)
-    assert isinstance(obj, bytes)
-
-    return identifier + obj
+                data = serializer.serialize(obj)
+                return identifier + b'\n' + data
+            except Exception:
+                continue
+    raise SerializationError(
+        f'Object of type {type(obj)} is not serializable.',
+    )
```

Comment:
I think it would be good to still raise the specific last exception that was caught.
```python
    last_exception: Exception | None = None
    for identifier, serializer in SERIALIZERS.items():
        if serializer.serializable(obj):
            try:
                data = serializer.serialize(obj)
                return identifier + b'\n' + data
            except Exception as e:
                last_exception = e

    assert last_exception is not None
    raise SerializationError(
        f'Object of type {type(obj)} is not serializable.',
    ) from last_exception
```